### PR TITLE
Change Route name to unique

### PR DESCRIPTION
**Pull request by nak3**
This patch changes reconciler of Route to:

- [x] use route name to unique by route's domain
- [x] clean up obsoleted route if found

### CODE REVIEW

File: pkg/controller/common/reconciler.go [hunk 1]
```diff
@@ -30,16 +30,32 @@ func (r *BaseIngressReconciler) ReconcileIngress(ctx context.Context, ci network
 
 	exposed := ci.GetSpec().Visibility == networkingv1alpha1.IngressVisibilityExternalIP
 	if exposed {
+		listOpts := &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				networking.IngressLabelKey: ci.GetName(),
```
Comment (markusthoemmes) [inline pkg/controller/common/reconciler.go]: I'd add the Knative route labels we add to the routes here as well as this is again prone to errors because it's not namespaced properly. For routes we add a proper namespace label though.

File: pkg/controller/common/reconciler.go [hunk 2]
```diff
@@ -30,16 +30,32 @@ func (r *BaseIngressReconciler) ReconcileIngress(ctx context.Context, ci network
...
 		}
-
 		for _, route := range routes {
 			logger.Infof("Creating/Updating OpenShift Route for host %s", route.Spec.Host)
 			if err := r.reconcileRoute(ctx, ci, route); err != nil {
 				return err
 			}
+			delete(existingMap, route.Name)
+		}
+		// If routes remains in existingMap, it must be obsoleted routes. Clean up them.
```
Comment (markusthoemmes) [inline pkg/controller/common/reconciler.go]:
```suggestion
		// If routes remains in existingMap, it must be obsoleted routes. Clean them up.
```

File: pkg/controller/common/reconciler.go [hunk 3]
```diff
@@ -30,16 +30,32 @@ func (r *BaseIngressReconciler) ReconcileIngress(ctx context.Context, ci network
...
+		for _, rt := range existingMap {
+			logger.Infof("Deleting obsoleted route: %s", rt.Name)
+			r.deleteRoute(ctx, rt)
```
Comment (markusthoemmes) [inline pkg/controller/common/reconciler.go]: Error handling?

File: pkg/controller/ingress/ingress_controller_test.go [hunk 4]
```diff
@@ -54,6 +55,63 @@ var (
...
 
+func TestRouteMigration(t *testing.T) {
+	logf.SetLogger(logf.ZapLogger(true))
+
+	tests := []struct {
+		name        string
+		exRoute     *routev1.Route
+		wantName    string
+		removedName string
+	}{
```
Comment (markusthoemmes) [inline pkg/controller/ingress/ingress_controller_test.go]:
```suggestion
	}{{
```

Spares you one level of indentation.

File: pkg/controller/ingress/ingress_controller_test.go [hunk 5]
```diff
@@ -54,6 +55,63 @@ var (
...
+		{
```
Comment (markusthoemmes) [inline pkg/controller/ingress/ingress_controller_test.go]: Why do a table test if we only have one test anyway?

File: pkg/controller/ingress/ingress_controller_test.go [hunk 6]
```diff
@@ -54,6 +55,63 @@ var (
 	}
 )
 
+func TestRouteMigration(t *testing.T) {
```
Comment (markusthoemmes) [inline pkg/controller/ingress/ingress_controller_test.go]: ❤️  Thank you for adding a test straight away 🙂

File: pkg/controller/common/reconciler.go [hunk 7]
```diff
@@ -49,8 +65,25 @@ func (r *BaseIngressReconciler) ReconcileIngress(ctx context.Context, ci network
 	return nil
 }
 
-func (r *BaseIngressReconciler) deleteRoutes(ctx context.Context, ci networkingv1alpha1.IngressAccessor) error {
+func routeMap(routes routev1.RouteList) map[string]*routev1.Route {
+	mp := map[string]*routev1.Route{}
```
Comment (markusthoemmes) [inline pkg/controller/common/reconciler.go]: nit: (Just to get into the habit of doing this)

```suggestion
	mp := make(map[string]*routev1.Route, len(routes.Items))
```

Allocates the map to the needed size immediately.

File: pkg/controller/resources/route.go [hunk 8]
```diff
@@ -131,7 +128,7 @@ func makeRoute(ci networkingv1alpha1.IngressAccessor, host string, index int, ru
 
 	route := &routev1.Route{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:            name,
+			Name:            host,
```
Comment (markusthoemmes) [inline pkg/controller/resources/route.go]: As discussed in Slack, I this *feels* a little off as we'll create a new set of routes when we change the domain config for example. I don't have a strong opinion though. @bbrowning do you?

File: pkg/controller/common/reconciler.go [hunk 9]
```diff
@@ -30,16 +31,37 @@ func (r *BaseIngressReconciler) ReconcileIngress(ctx context.Context, ci network
...
 			if err := r.reconcileRoute(ctx, ci, route); err != nil {
 				return err
 			}
+			delete(existingMap, route.Name)
+		}
+		// If routes remains in existingMap, it must be obsoleted routes. Clean them up.
+		for _, rt := range existingMap {
+			logger.Infof("Deleting obsoleted route: %s", rt.Name)
+			if err := r.deleteRoute(ctx, rt); err != nil {
+				logger.Warnf("Failed to delete obsoleted route %s: %v", rt.Name, err)
```
Comment (markusthoemmes) [inline pkg/controller/common/reconciler.go]: Why not actually return the error here? I guess the same goes for line 67.

File: pkg/controller/ingress/ingress_controller_test.go [hunk 10]
```diff
@@ -54,6 +55,60 @@ var (
...
+
+	test := struct {
+		name        string
+		exRoute     *routev1.Route
+		wantName    string
+		removedName string
+	}{
+		name: "Clean up old route and new route is generated",
+		exRoute: &routev1.Route{
+			ObjectMeta: metav1.ObjectMeta{Name: "test-0"},
```
Comment (markusthoemmes) [inline pkg/controller/ingress/ingress_controller_test.go]: should this have the correct labels set? Could we add a second route here that has different labels and make sure we only remove the one that matches?

File: pkg/controller/resources/route.go (hunk 8 above)
Comment (markusthoemmes) [reply pkg/controller/resources/route.go]: Discussed this with @bbrowning, who was uncomfortable with just the hostname as a unique thing as well. I also realized that `$ns-$name-$index` isn't necessarily unique (ex: ns: test, name: test-service and ns: test-test, name service would result in the same thing).

Shall we use `route-${ci.GetUID()}-$index` instead? that should be guaranteed to be unique. Not exactly user-friendly but the user isn't supposed to mess with this anyway.

WDYT @nak3 @savitaashture ?

File: pkg/controller/ingress/ingress_controller_test.go (hunk 10 above)
Comment (nak3) [reply pkg/controller/ingress/ingress_controller_test.go]: > Could we add a second route here that has different labels and make sure we only remove the one that matches?

Yes, we should add it. But it looks like __fake__ client in current K-O-I does not support labels list option added by https://github.com/kubernetes-sigs/controller-runtime/commit/e9c5505ea8a3bf1bb77b5f15a9fd904c597e63e6, so we cannot test it for now.
We need to upgrade the controller-runtime version for the test.

File: pkg/controller/resources/route.go (hunk 8 above)
Comment (nak3) [reply pkg/controller/resources/route.go]: It would be fine for uniqueness. But as reported https://github.com/openshift-knative/knative-openshift-ingress/pull/16, it would be more difficult to read the log message...

File: pkg/controller/common/reconciler.go [hunk 11]
```diff
@@ -30,16 +31,37 @@ func (r *BaseIngressReconciler) ReconcileIngress(ctx context.Context, ci network
...
+
 		routes, err := resources.MakeRoutes(ci)
 		if err != nil {
 			return err
 		}
-
 		for _, route := range routes {
 			logger.Infof("Creating/Updating OpenShift Route for host %s", route.Spec.Host)
 			if err := r.reconcileRoute(ctx, ci, route); err != nil {
 				return err
```
Comment (markusthoemmes) [inline pkg/controller/common/reconciler.go]: To fix concerns regarding logging, let's augment this a little bit to aid in readability:

```suggestion
				return fmt.Errorf("failed to create route for host %s: %v", route.Spec.Host, err)
```

File: pkg/controller/ingress/ingress_controller_test.go (hunk 10 above)
Comment (markusthoemmes) [reply pkg/controller/ingress/ingress_controller_test.go]: ugh. Can we safely do that? A separate PR first maybe?

File: pkg/controller/resources/route.go (hunk 8 above)
Comment (markusthoemmes) [reply pkg/controller/resources/route.go]: Oh that we can fix differently. Commented on the relevant lines to add more context to our logs.

File: pkg/controller/common/reconciler.go [hunk 12]
```diff
@@ -49,8 +71,25 @@ func (r *BaseIngressReconciler) ReconcileIngress(ctx context.Context, ci network
...
+		mp[route.Name] = &route
+	}
+	return mp
+}
+
+func (r *BaseIngressReconciler) deleteRoute(ctx context.Context, route *routev1.Route) error {
 	logger := logging.FromContext(ctx)
+	logger.Infof("Deleting OpenShift Route for host %s", route.Spec.Host)
+	if err := r.Client.Delete(ctx, route); err != nil {
+		return err
```
Comment (markusthoemmes) [inline pkg/controller/common/reconciler.go]: Let's add the same context to the error as above here for consistency and easier debugability.

File: pkg/controller/ingress/ingress_controller_test.go (hunk 10 above)
Comment (markusthoemmes) [reply pkg/controller/ingress/ingress_controller_test.go]: I did check the possibility of doing so quickly:
- @nak3's referenced commit is part of controller-runtime since v0.2.0.
- To get that, we'd need to upgrade to osdk v0.11.0
- osdk v0.11.0 also bumped k8s clients to `kubernetes-1.14.1` though, but knative-serving wants `kubernetes-1.13.1`.

It might be more trouble than it's worth it at this point. I'd personally be fine if we just check this manually for this release and make sure it does the right thing. Alternatively, could we implement the label matching in our code, leaving a comment that we do this in addition to the List filter so that we can test things?

**Comment by nak3** [comment]
/hold

I still need to address https://github.com/openshift-knative/knative-openshift-ingress/pull/24#discussion_r341804021

File: pkg/controller/ingress/ingress_controller_test.go [hunk 13]
```diff
@@ -62,18 +64,35 @@ func TestRouteMigration(t *testing.T) {
...
 
 	test := struct {
-		name        string
-		exRoute     *routev1.Route
-		wantName    string
-		removedName string
+		name            string
+		exRoute         *routev1.Route
+		noLabelRoute    *routev1.Route
+		otherLabelRoute *routev1.Route
```
Comment (markusthoemmes) [inline pkg/controller/ingress/ingress_controller_test.go]: nit: Usually, these tests are written with a „state“ slice of entities and a „want“ slice of entities that we test against.

In this specific case we‘d pass all existing routes as „state“, call Reconcile, list all routes again and check what we‘ve got against the „want“ slice. Testify also has nice functions to assert that the resulting list has the same elements as the „want“ list.

That makes the test more generally useful and easier to extent with more edge cases.

File: pkg/controller/ingress/ingress_controller_test.go [hunk 14]
```diff
@@ -54,6 +60,88 @@ var (
...
+			NamespacedName: types.NamespacedName{
+				Name:      name,
+				Namespace: namespace,
+			},
+		}
+		if _, err := r.Reconcile(req); err != nil {
+			t.Fatalf("reconcile: (%v)", err)
+		}
+
+		route := &routev1.Route{}
```
Comment (markusthoemmes) [inline pkg/controller/ingress/ingress_controller_test.go]: Per the comment above, we should be able to replace all of this via a check against a „want“ list of routes

File: pkg/controller/ingress/ingress_controller_test.go [hunk 15]
```diff
@@ -54,6 +63,104 @@ var (
...
+		if _, err := r.Reconcile(req); err != nil {
+			t.Fatalf("reconcile: (%v)", err)
+		}
+
+		routeList := &routev1.RouteList{}
+		err := cl.List(context.TODO(), &client.ListOptions{}, routeList)
+		assert.Nil(t, err)
+
+		routes := routeList.Items
+		assert.Equal(t, len(test.want), len(routes))
```
Comment (markusthoemmes) [inline pkg/controller/ingress/ingress_controller_test.go]: Thanks, this is soooo much better 🙂 . I'd suggest to even use the `ElementsMatch` function to be able to remove all of the manual sorting and checking. Possible diff here: https://github.com/openshift-knative/knative-openshift-ingress/commit/c729300e613d20ed71b8599edee3f90c3c8dfdd1

Why?
- It makes the test a lot simpler and directly expresses the intent. 
- It also makes sure to check the entire objects, which is sometimes annoying to setup but almost always worth the hassle in terms of test coverage.
- Most importantly: It produces readable debug outcome. If this fails, it'll say something like "3 was not 2" which tells you nothing to debug. `ElementsMatch` will output a diff that you can use to see what's going one.

**Comment by nak3** [comment]
Sure, thank you so much for your review! It becomes significantly better.

> Last but not least: This PR will need to get sent against the 0.0.x branch as well, 0.9 will be built from that. Thanks!

Yes, I will send against the other branches after passed the review.

File: pkg/controller/ingress/ingress_controller_test.go [hunk 16]
```diff
@@ -54,6 +63,101 @@ var (
...
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   "no-remove-other-label",
+				Labels: map[string]string{networking.IngressLabelKey: "another", serving.RouteLabelKey: name, serving.RouteNamespaceLabelKey: namespace},
+			},
+			Spec: routev1.RouteSpec{Host: "c.example.com"},
+		}},
+		want: []routev1.Route{{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            routeName0,
+				Namespace:       "istio-system",
```
Comment (savitaashture) [inline pkg/controller/ingress/ingress_controller_test.go]: nit
```suggestion
				Namespace:      namespace,
```

**Comment by savitaashture** [comment]
@nak3 Thank you for the changes :+1:
